### ISSUE TITLE
Cryptography misuse

### ISSUE THREAD

**Issue by z0mbie42** [Q-Vault/qvault#311]
Hi,
I noticed the following cryptography misuse:

* AES GCM requires a 12 bytes IV , not 16 (https://crypto.stackexchange.com/questions/41601/aes-gcm-recommended-iv-size-why-12-bytes/41610) -> https://github.com/Q-Vault/qvault/blob/master/src/lib/QVaultCrypto/QVaultCrypto.js#L193
* AES GCM provides authentication, it allows to assure data integrity (ex: encrypted password have not been swapped...). Thus you SHOULD use AES GCM with Additional Authenticated Data (https://nodejs.org/api/crypto.html#crypto_cipher_setaad_buffer_options), like the site url of the credentials
* ~Regarding the Master key generation, I think you should use an appropriate Password-Based Key Derivation Function instead of a simple truncated hash, like PBKDF2~

There may be other cryptography misuse, it's just the result of a quick overview.

**Comment by lane-c-wagner** [comment]
Hey, thanks for bringing these up! I'll look into integrating your first 2 suggestions in a backwards-compatible way. 
Regarding your third point, we are using the Scrypt KDF, unless you are talking about something else?

**Comment by z0mbie42** [comment]
Yes my bad,

On the other hand, I'm not sure to understand
https://github.com/Q-Vault/qvault/blob/7e70908089cc2cb0e8ad3d950b5b1c400279fbcf/src/lib/QVaultCrypto/QVaultCrypto.js#L192

as the key should already be 32 bytes from
https://github.com/Q-Vault/qvault/blob/7e70908089cc2cb0e8ad3d950b5b1c400279fbcf/src/lib/QVaultCrypto/QVaultCrypto.js#L179

Is there any manipulation of the key after the hashing? (which may reduce entropy/security)

**Comment by lane-c-wagner** [comment]
No there isn't, the important part is converting from string to byteArray. I'll look into the .slice() part thought, tbh I can't remember why we needed that (maybe we don't need it anymore)

**Comment by z0mbie42** [comment]
Another thing:
In a multi devices/cloud environment what you want is to HMAC the vault, and not simply hash it's content.

The reason is ghat with a simple hash the server can modify the vault whereas with HMAC a secret (derived from password, or generated and then encrypted) is required to produce a valid signature

**Comment by lane-c-wagner** [comment]
I think you may be misunderstanding the purpose of the cloud_vault_hash. It isn't a security measure, because that wouldn't offer any security anyhow. That hash only exists so that when a user goes to save a vault, the software can know that it won't be overwriting any changes that it wasn't aware of. 

For example, if I made an update on my laptop, then make an update on my desktop and for some reason I never downloaded the laptop changes, I don't want to update my cloud version.

**Comment by z0mbie42** [comment]
Ok, I thought it was for vault integrity check.

From my point of view, the threat model for a cloud based password manager is to not trust the server. Because hashes are not authenticated, it looks like in your scenario a bad server can alter the vault.

By using HMAC, it would allow the 'update check' and also in the case the vault is downloaded, to assert  it's provenance (integrity, from a device which has the HMAC secret key).

**Comment by lane-c-wagner** [comment]
I agree that the server shouldn't be trusted. The server doesn't have access to the key used to encrypt the vault, so if it tried to swap vaults the user wouldn't be able to decrypt. This will be even better once we have full GCM capabilities like you describe above.

I don't think an hmac is particularly useful here because there are legitimate uses where the user is downloading their vault to a new device that has no way to authenticate other than by trying to decrypt.

Am I understanding correctly?

**Comment by z0mbie42** [comment]
Hi,

If the entire vault is encrypted, AES-GCM should be enough to authenticate the encrypted vault, but if some metadata is hashed with the vault then HMAC should be preferred.

**Comment by lane-c-wagner** [comment]
So there are really 3 issues you brought up in this thread that are still unresolved. I hope to resolve some of them here.

1. Using an IV of length 16 vs 12:
 I read through the stackexchange post you listed and it seems that the consensus view was that there is no security loss by using 16 bytes, just some potential performance loss, which I'm not worried about. I added a comment explaining the situation in the code so that if it ever comes up again we have context.

2. Adding AAD for authentication:
After looking into it more (see https://cloud.google.com/kms/docs/additional-authenticated-data) It becomes clear that AAD is built for a system where the attacker will have difficulty finding the additional data (like a server environment). Because Qvault isn't used in that way, the authTag that we are currently using seems the most effective and all that is necessary.

3. I pushed a fix to the .slice(0,32) weirdness in the above PR
